### PR TITLE
  Support daemonized execution

### PR DESCRIPTION
**Pull request by aabawer**
Examples for daemonized executions:
    
    ./ovirt-ci   --daemon  system-tests   99762
    ./ovirt-ci   --daemon  build-artifacts   99762
    
    ovirt-ci will execute in background and output will be redirected
    to a log file named /var/tmp/oci_{command}_{change id}.log

### CODE REVIEW

File: oci/main.py [hunk 1]
```diff
@@ -16,12 +17,17 @@ def run():
         '--debug',
         help="Show noisy debug logs",
         action="store_true")
+
+    parser.add_argument(
+        '--daemon',
+        help="Run as daemon and redirect output to /var/tmp/oci_{command}_{change id}.log",
+        action="store_true")
```
Comment (nirs) [inline oci/main.py:24]: This option is reasonable, but using /var/tmp/oci_... will create lot of junk logs that someone will have to cleanup. This is a usability problem.

Also using change id, later builds will override older builds logs, hiding errors you may want to inspect.

File: oci/main.py [hunk 2]
```diff
@@ -112,6 +125,13 @@ def system_tests(args):
 
     out.success("System tests completed successfully, congratulations!")
 
+def invoke_build_artifacts(args):
+    if args.daemon:
+        with open("/var/tmp/oci_build_artifcats_{}_.log".format(args.change), 'w') as f:
+            with daemon.DaemonContext(stdout=f,stderr=f):
+                build_artifacts(args)
+    else:
+        build_artifacts(args)
```
Comment (nirs) [inline oci/main.py:134]: This is poor way to integrate such change. It means that we need to duplicate this boilerplate for any new command.

If we have code daemonizing, it should run after parsing arguments in run(), before we invoke the command functions.

**Review by nirs — COMMENTED** [review]
I don't think this is a good idea to implement daemon option for command line tool.

This creates many usability problems:
- How do you know when your build has finished or failed? When running in the foreground, the progress and errors are visible, and you get notifications when the command ends (at least on Gnome on Fedora).
- How do you clean up all the logs? This creates extra work for the user
- How do you find all the running instances and terminate them? when you run in the forground, it is very easy to close a window/tab or interrupt the tool
- If you run 2 builds in the same time (sometimes builds are flaky, running 2 builds can save you 40 minutes of wait), they will overwrite each other log.
- And the most important issue, why would someone want to run a build in the background, instead of in a new tab or shell?

There are many solutions to run a build or several builds at the same time - without changing anything in the code:
- Open new terminal tab or window, run the command
- If you want to run the tool on a remote machine and make the tool protected from hangup, use nohup: (nohup ./ovirt-ci system-tests 12345)
- More advanced user may want to use screen or tmux to run many commands at the same time, protected from hangups

In summary I don't think this feature is needed, but if you want to contribute to this project please check the current issues, there are many interesting challenges like #11, #28, #13, and #12.

If you think not having a daemon option please file an issue and describe why this feature is important.

**Comment by aabawer** [comment]
> I don't think this is a good idea to implement daemon option for command line tool.
> 
> This creates many usability problems:
> 
> * How do you know when your build has finished or failed? When running in the foreground, the progress and errors are visible, and you get notifications when the command ends (at least on Gnome on Fedora).
ANS: tail -F /var/tmp/logfile
> * How do you clean up all the logs? This creates extra work for the user
ANS: logrotate
> * How do you find all the running instances and terminate them? when you run in the forground, it is very easy to close a window/tab or interrupt the tool
ANS: sudo killall ovirt-ci
> * If you run 2 builds in the same time (sometimes builds are flaky, running 2 builds can save you 40 minutes of wait), they will overwrite each other log.
ANS: can add execution timestamp to logname, aside to command and changeid
> * And the most important issue, why would someone want to run a build in the background, instead of in a new tab or shell?
ANS: @barpavel
> 
> There are many solutions to run a build or several builds at the same time - without changing anything in the code:
> 
> * Open new terminal tab or window, run the command
> * If you want to run the tool on a remote machine and make the tool protected from hangup, use nohup: (nohup ./ovirt-ci system-tests 12345)
> * More advanced user may want to use screen or tmux to run many commands at the same time, protected from hangups
> 
> In summary I don't think this feature is needed, but if you want to contribute to this project please check the current issues, there are many interesting challenges like #11, #28, #13, and #12.
> 
> If you think not having a daemon option please file an issue and describe why this feature is important.
ANS: i think we should refer this question to the potential users :)

File: oci/main.py (hunk 1 above)
Comment (aabawer) [reply oci/main.py:24]: can add timestamp to file name as well

File: oci/main.py (hunk 2 above)
Comment (aabawer) [reply oci/main.py:134]: can add boilerplate
def executer(cmd, args)
which will run cmd with args according to arg.daemon

**Comment by nirs** [comment]
If you want to run commands in the background, and the current tools to do that are not good enough, why not create a tool that will run any command in the background?

**Comment by nirs** [comment]
Closing old PRs. Feel free to reopen if you think this is still needed.
